### PR TITLE
Recursively expand directories before delete and enable folder rename functional tests

### LINKED ISSUE #202: Mac: Folders should be expanded before being deleted
The first step of a folder rename is a PreDelete notification, and when handling this notification we must expand the folder (and all of its children) to ensure all children are on disk prior to the rename proceeding.

### PR DESCRIPTION
**Pull request by wilbaker**
Fixes #202

Summary of changes:
- Added a new kernel->user message `MessageType_KtoU_RecursivelyEnumerateDirectory`
- The new message is sent from the kext to user mode whenever a directory delete is about to occur
- Handling of the message is done by the user mode library.  The lib walks the directory and asks the provider to expands any empty directories it finds
- Enabled partial folder rename functional tests

### CODE REVIEW

File: ProjFS.Mac/PrjFSLib/PrjFSLib.cpp [hunk 1]
```diff
@@ -588,6 +597,68 @@ static PrjFS_Result HandleEnumerateDirectoryRequest(const MessageHeader* request
...
+    directoryRelativePaths.push_back(path);
+    
+    // Walk each directory, expanding those that are found to be empty
+    char pathBuffer[PrjFSMaxPath];
+    std::list<std::string>::iterator relativePathIter = directoryRelativePaths.begin();
+    while(relativePathIter != directoryRelativePaths.end())
+    {
+        CombinePaths(s_virtualizationRootFullPath.c_str(), relativePathIter->c_str(), pathBuffer);
+    
+        // TODO(Mac): how should we handle scenarios where we were unable to fully expand the directory and
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: Thinking about this a bit more, I'm actually not sure we have any new partial failure modes.

With this change we're still expanding folders one at a time, and if any of those expansions fails we stop attempting to expand.

I will confirm with @nickgra but I don't think this will change how the provider expands directories.  It would be no different from a user manually expanding several folders (one at a time).

The new failure mode we are introducing is the kext blocking the delete of an empty folder if the provider is not able to expand it.

[the remark this answers is not in the document]
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: > what will happen in the scenario where a directory's recursive children are partially on disk already

In your example, assuming we get `KAUTH_VNODE_DELETE` for `A` first, the directories would be walked breadth-first:

1. Push `A` into `directoryRelativePaths`
2. Grab the next entry in `directoryRelativePaths` (i.e. `A`), it does not need to be expanded
3. Walk all children of `A` and push any directories (i.e. `B` and `C`) onto `directoryRelativePaths`
4. Grab the next entry in `directoryRelativePaths` (i.e. `B`), it does not need to be expanded
5. Walk all children of `B`, there are no directories
6. Grab the next entry in `directoryRelativePaths` (i.e. `C`), it needs to be expanded, and so expand it
7. Walk all children of `C`, there are no directories
8. There are no more entries in `directoryRelativePaths`

(Note that writing this out I realize we should probably just pull the entry off the front of `directoryRelativePaths` after we process it rather than leaving it in `directoryRelativePaths`)

> How does that behave normally?

It's possible that without my changes we'd already be expanding everything (because the delete is recursive).  I can test that out and confirm.  

The real reason we need this change is for rename operations, as they will not walk the tree recursively.

> How does it behave if some of the expansions fail, especially with more complicated data shapes than my example?

I will test out delete specifically and see in what order `KAUTH_VNODE_DELETE` is received.

At the end of the day though, if any of the expansions fail the kext will return `DENY` and prevent the delete\rename from proceeding.  I can force some failures to occur and reply back with what I see.

Assuming that each subfolder isn't marked as expanded until all of its placeholders are written, then (as mentioned in the earlier comment), these changes aren't introducing much new complexity.  If the delete is requested again the user-mode library will skip over the folders that were expanded in the previous attempt.

**Comment by wilbaker** [comment]
@sanoursa @nickgra this is ready for another look, let me know what you think.

File: ProjFS.Mac/PrjFSLib/PrjFSLib.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nickgra) [reply ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: I agree, I don't think this introduces any new partial failure modes. 

If one of these expansions fails, the state in the virtualized path will be in a state where it can be repaired by retrying the I/O, even if we fail in the middle of a directory enumeration. 

From the design for atomic placeholders:
"In a case where we crash (or the machine crashes, or the user kills the mount process) while expanding a placeholder directory, we will be able to recover by calling UpdatePlaceholderIfNeeded on files that were enumerated by the previous attempt and creating the rest of the placeholders as expected."

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/MoveRenameFolderTests.cs [hunk 2]
```diff
@@ -181,14 +174,12 @@ public void MoveAndRenameUnhydratedFolderToNewFolder()
...
 
-        // This test requires expansion on PreDelete to be implemented
         // MacOnly because renames of partial folders are blocked on Windows
         [TestCase]
-        [Category(Categories.MacTODO.M2)]
         [Category(Categories.MacOnly)]
         public void MoveFolderWithUnhydratedAndFullContents()
         {
-            string testFileName = "MoveFolderWithUnhydratedAndFullContents.cs";
+            string testFileName = "MoveFolderWithUnhydratedAndFullContents.cpp";
```
Comment (nickgra) [inline GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/MoveRenameFolderTests.cs]: Why was this changed?

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp [hunk 3]
```diff
@@ -288,6 +289,22 @@ static int HandleVnodeOperation(
         goto CleanupAndReturn;
     }
     
+    if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+    {
+        if (!TrySendRequestAndWaitForResponse(
+                root,
+                VDIR == vnodeType ? MessageType_KtoU_NotifyDirectoryPreDelete : MessageType_KtoU_NotifyFilePreDelete,
```
Comment (nickgra) [inline ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: This TrySendRequestAndWaitForResponse and the one below structure their ternary statements differently. I have a preference for how I think it should look, but you can ignore that so long as they're consistent with each other :)

Can we pull the comparison up into a separate variable here too? Additionally, this line is getting long, I'd rather we break it up into 3 separate lines like we do in the below addition.

File: ProjFS.Mac/PrjFSLib/PrjFSLib.cpp [hunk 4]
```diff
@@ -588,6 +597,68 @@ static PrjFS_Result HandleEnumerateDirectoryRequest(const MessageHeader* request
     return callbackResult;
 }
 
+static PrjFS_Result HandleRecursivelyEnumerateDirectoryRequest(const MessageHeader* request, const char* path)
+{
+#ifdef DEBUG
+    std::cout << "PrjFSLib.HandleRecursivelyEnumerateDirectoryRequest: " << path << std::endl;
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: > Is this deliberate, or is it a case of not getting around to putting a bunch of using statements at the top yet so far?

Several years ago I picked up the habit of always using full namespaces in C++ as I'd been burned by poor tool support when performing large scale refactoring.  At that time I was working on code where class names existed in several namespaces and without the full namespace it was painful to track down all the uses of a particular class.

I'm happy to switch to `using` though if that's the team preference :smile:

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/MoveRenameFolderTests.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/MoveRenameFolderTests.cs]: The old path was just wrong (see [MoveFolderWithUnhydratedAndFullContents.cpp](https://gvfs.visualstudio.com/ci/_git/ForTests?path=%2FTest_EPF_MoveRenameFolderTests%2FMoveFolderWithUnhydratedAndFullContents%2FMoveFolderWithUnhydratedAndFullContents.cpp&version=GBFunctionalTests%2F20180214)).

This hasn't caused any failures up until now as the tests have been disabled on Windows for a very long time.

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 3 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: > This TrySendRequestAndWaitForResponse and the one below structure their ternary statements differently. I have a preference for how I think it should look, but you can ignore that so long as they're consistent with each other

I can make them consistent, but could you provide more details on how you see them as inconsistent?

> Can we pull the comparison up into a separate variable here too? Additionally, this line is getting long, I'd rather we break it up into 3 separate lines like we do in the below addition.

Yes to both of these, I'll make the change.

File: ProjFS.Mac/PrjFSLib/PrjFSLib.cpp [hunk 5]
```diff
@@ -7,6 +7,9 @@
 #include <sys/xattr.h>
 #include <thread>
 #include <unistd.h>
+#include <dirent.h>
+#include <list>
```
Comment (wilbaker) [inline ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: Remove `#include <list>`

File: ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp (hunk 3 above)
Comment (wilbaker) [reply ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp]: @nickgra I think I've addressed this feedback with my latest commit, but let me know if there's still changes you'd like to see.

File: ProjFS.Mac/PrjFSLib/PrjFSLib.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: Actually, I see now that this file a bit of an inconsistent mess.  I will clean this up (to use `using`) and thanks again for the feedback.

[the remark this answers is not in the document]
Comment (wilbaker) [reply ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: >  At that time I was working on code where class names existed in several namespaces and without the full namespace it was painful to track down all the uses of a particular class.

And I see that we're not introducing the entire `std` namespace here so we avoid this issue as well :)
